daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -95,28 +61,52 @@ private[preprocessing] final class TransactionPreprocessor(
       tx: GenTransaction[NodeId, Cid]
   ): (ImmArray[speedy.Command], Set[ContractId]) = {
 
-    type Acc = ((Set[Value.ContractId], Set[Value.ContractId]), BackStack[speedy.Command])
+    type Acc = (Set[Value.ContractId], Set[Value.ContractId], BackStack[speedy.Command])
 
-    val ((localCids, _), cmds) =
-      tx.roots.foldLeft[Acc](((Set.empty, Set.empty), BackStack.empty)) {
-        case ((cids, stack), id) =>
-          tx.nodes.get(id) match {
+    val (_, globalCids, cmds) =
+      tx.roots.foldLeft[Acc]((Set.empty, Set.empty, BackStack.empty)) {
+        case ((localCids0, globalCids0, stack), id) =>
+          val (newGlobals, newLocals, cmd) = tx.nodes.get(id) match {
             case None =>
               fail(s"invalid transaction, root refers to non-existing node $id")
             case Some(node) =>
               node match {
+                case create: Node.NodeCreate[Cid] =>
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessCreate(create.templateId, create.arg)
+                  val newGlobalCids = newCids.filterNot(localCids0)
+                  (newGlobalCids, List(create.coid), cmd)
+                case exe: Node.NodeExercises[_, Cid] =>
+                  val templateId = exe.templateId
```

Comment:
nit: seems odd, that this is the only thing that you pull out into a local variable and then you still only use it once.

Reply:
Not sure what you’re getting at. All I was suggesting is to inline this or either assign all the `exe._` fields to variables.

Reply:
Of course. thanks.